### ISSUE TITLE
[bug] Package using apply_conandata_patches() sometimes throws exception when built in both Build and Host profile

### ISSUE THREAD

**Issue by je894591** [conan-io/conan#11854]
### Environment Details
  * Operating System+version: Debian Buster
  * Compiler+version: GCC 8.3  on Debian
  * Conan version: 1.51.2
  * Python version: 3.7

I've used zlib/1.2.12 as a reproducer because it was recently updated to use `conan.tools.files.apply_conandata_patches()` and started causing internal build failures.  This particular package ends up getting built twice by our consumer conanfile.py: once in the host context due to a direct dependency, plus a second time in the build context due to a protobuf tool_requires that pulls in the library again.

A minimal Debian conanfile.txt reproducer for this bug report is attached: [reproducer.tar.gz](https://github.com/conan-io/conan/files/9312899/reproducer.tar.gz)

I also reproduced this on macOS with Python 3.9 and the apple-clang compiler provided by Xcode 13.4.1.

### Steps to Reproduce

Create a conanfile.txt with the zlib/1.2.12 declared in both the build_requires and tool_requires section, then define separate build and host profiles as per the [cross building docs][cross-build]. The separate architectures require two different package builds:

```
conanfile.txt: Installing package
Requirements
    zlib/1.2.12 from 'conancenter' - Cache
Packages
    zlib/1.2.12:9f5d845a9acd4104ba44865c80f5b042812f136e - Build
Build requirements
    zlib/1.2.12 from 'conancenter' - Cache
Build requirements packages
    zlib/1.2.12:1cfd15a0464c4b1c680c61895450fbdbe189331c - Build
```

[cross-build]: https://docs.conan.io/en/latest/systems_cross_building/cross_building.html

Invoke `conan install` with `--build "zlib"` to force the zlib package to build twice: once in the host profile, which will always pass, and then a second time in the build profile. This second build dies on an exception thrown by `conan.tools.apply_conandata_packages`:

```
ERROR: zlib/1.2.12: Error in build() method, line 86
	self._patch_sources()
while calling '_patch_sources', line 70
	apply_conandata_patches(self)
	ConanException: The 'conandata.yml' file needs a 'patch_file' or 'patch_string' entry for every patch to be applied
```

Inserting a try-catch and breakpoint into the exported zlib recipe shows that `self.conan_data` has corrupted patch data (this might just a missing deep copy when instantiating the second ZlibConanfile instance from the catch?):

```
> /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/export/conanfile.py(74)_patch_sources()
-> pass
(Pdb) list
 69  	    def _patch_sources(self):
 70  	        try:
 71  	            apply_conandata_patches(self)
 72  	        except Exception as e:
 73  	            import pdb; pdb.set_trace()
 74  ->	            pass
 75  	
 76  	        is_apple_clang12 = self.settings.compiler == "apple-clang" and Version(self.settings.compiler.version) >= "12.0"
 77  	        if not is_apple_clang12:
 78  	            for filename in ['zconf.h', 'zconf.h.cmakein', 'zconf.h.in']:
 79  	                filepath = os.path.join(self.source_folder, filename)
(Pdb) e
ConanException("The 'conandata.yml' file needs a 'patch_file' or 'patch_string' entry for every patch to be applied")
(Pdb) self.conan_data
{'patches': {'1.2.12': [{}, {}]}, 'sources': {'1.2.12': {'sha256': '91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9', 'url': 'https://zlib.net/fossils/zlib-1.2.12.tar.gz'}}}
```

For reference, zlib/1.2.12 should have two patch files defined:

```
❯ cat .conan/data/zlib/1.2.12/_/_/export/conandata.yml 
patches:
  1.2.12:
  - patch_file: patches/0001-fix-cmake.patch
  - patch_file: patches/0002-gzguts-xcode12-compile-fix.patch
sources:
  1.2.12:
    sha256: 91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9
    url: https://zlib.net/fossils/zlib-1.2.12.tar.gz
```

### Logs

<details>
<summary>Full Output (always fails due to --build "zlib")</summary>

```
$ CONAN_USER_HOME="$(pwd)" conan install --profile:host host-x86_64 --profile:build build-aarch64 --build "zlib" conanfile.txt
Configuration (profile_host):
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8.3
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
AR=/usr/x86_64-linux-gnu/bin/ar
AS=/usr/x86_64-linux-gnu/bin/as
CC=x86_64-linux-gnu-gcc
CHOST=x86_64-linux-gnu
CXX=x86_64-linux-gnu-g++
LD=/usr/x86_64-linux-gnu/bin/ld
PATH=[/usr/x86_64-linux-gnu/bin]
RANLIB=/usr/x86_64-linux-gnu/bin/ranlib
STRIP=/usr/x86_64-linux-gnu/bin/strip
Configuration (profile_build):
[settings]
arch=armv8
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8.3
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

zlib/1.2.12: Forced build from source
zlib/1.2.12: Forced build from source
conanfile.txt: Installing package
Requirements
    zlib/1.2.12 from 'conancenter' - Cache
Packages
    zlib/1.2.12:9f5d845a9acd4104ba44865c80f5b042812f136e - Build
Build requirements
    zlib/1.2.12 from 'conancenter' - Cache
Build requirements packages
    zlib/1.2.12:1cfd15a0464c4b1c680c61895450fbdbe189331c - Build

Cross-build from 'Linux:armv8' to 'Linux:x86_64'
Installing (downloading, building) binaries...
zlib/1.2.12: Copying sources to build folder
zlib/1.2.12: Building your package in /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c
zlib/1.2.12: Generator txt created conanbuildinfo.txt
zlib/1.2.12: Calling generate()
zlib/1.2.12: Aggregating env generators
zlib/1.2.12: Calling build()
zlib/1.2.12: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/src"
-- Using Conan toolchain: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is GNU 8.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - done
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Renaming
--     /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/src/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/Release
zlib/1.2.12: CMake command: cmake --build "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/Release" '--' '-j6'
Scanning dependencies of target zlib
[ 12%] Building C object CMakeFiles/zlib.dir/compress.c.o
[ 12%] Building C object CMakeFiles/zlib.dir/adler32.c.o
[ 18%] Building C object CMakeFiles/zlib.dir/crc32.c.o
[ 25%] Building C object CMakeFiles/zlib.dir/gzclose.c.o
[ 31%] Building C object CMakeFiles/zlib.dir/deflate.c.o
[ 37%] Building C object CMakeFiles/zlib.dir/gzlib.c.o
[ 43%] Building C object CMakeFiles/zlib.dir/gzread.c.o
[ 50%] Building C object CMakeFiles/zlib.dir/gzwrite.c.o
[ 56%] Building C object CMakeFiles/zlib.dir/inflate.c.o
[ 62%] Building C object CMakeFiles/zlib.dir/infback.c.o
[ 68%] Building C object CMakeFiles/zlib.dir/inftrees.c.o
[ 75%] Building C object CMakeFiles/zlib.dir/inffast.c.o
[ 81%] Building C object CMakeFiles/zlib.dir/trees.c.o
[ 87%] Building C object CMakeFiles/zlib.dir/uncompr.c.o
[ 93%] Building C object CMakeFiles/zlib.dir/zutil.c.o
[100%] Linking C static library libz.a
[100%] Built target zlib
zlib/1.2.12: Package '1cfd15a0464c4b1c680c61895450fbdbe189331c' built
zlib/1.2.12: Build folder /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/Release
zlib/1.2.12: Generated conaninfo.txt
zlib/1.2.12: Generated conanbuildinfo.txt
zlib/1.2.12: Generating the package
zlib/1.2.12: Package folder /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c
zlib/1.2.12: Calling package()
zlib/1.2.12: CMake command: cmake --install "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c/build/Release" --prefix "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c"
-- Install configuration: "Release"
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c/lib/libz.a
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c/include/zconf.h
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/1cfd15a0464c4b1c680c61895450fbdbe189331c/include/zlib.h
zlib/1.2.12 package(): Packaged 1 file: LICENSE
zlib/1.2.12 package(): Packaged 2 '.h' files: zconf.h, zlib.h
zlib/1.2.12 package(): Packaged 1 '.a' file: libz.a
zlib/1.2.12: Package '1cfd15a0464c4b1c680c61895450fbdbe189331c' created
zlib/1.2.12: Created package revision 2c83270bc4ee6758666b63dd61b57b7d
zlib/1.2.12: Copying sources to build folder
zlib/1.2.12: Building your package in /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e
zlib/1.2.12: Generator txt created conanbuildinfo.txt
zlib/1.2.12: Calling generate()
zlib/1.2.12: Aggregating env generators
zlib/1.2.12: Calling build()
zlib/1.2.12: 
zlib/1.2.12: WARN: Build folder is dirty, removing it: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/1cfd15a0464c4b1c680c61895450fbdbe189331c
zlib/1.2.12: ERROR: Package '9f5d845a9acd4104ba44865c80f5b042812f136e' build failed
zlib/1.2.12: WARN: Build folder /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/Release
ERROR: zlib/1.2.12: Error in build() method, line 86
	self._patch_sources()
while calling '_patch_sources', line 70
	apply_conandata_patches(self)
	ConanException: The 'conandata.yml' file needs a 'patch_file' or 'patch_string' entry for every patch to be applied
```
</details>

<details>
<summary>Passing Second Build (--build "missing")</summary>

```
❯ CONAN_USER_HOME="$(pwd)" conan install --profile:host host-x86_64 --profile:build build-aarch64 --build "missing" conanfile.txt
Configuration (profile_host):
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8.3
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
AR=/usr/x86_64-linux-gnu/bin/ar
AS=/usr/x86_64-linux-gnu/bin/as
CC=x86_64-linux-gnu-gcc
CHOST=x86_64-linux-gnu
CXX=x86_64-linux-gnu-g++
LD=/usr/x86_64-linux-gnu/bin/ld
PATH=[/usr/x86_64-linux-gnu/bin]
RANLIB=/usr/x86_64-linux-gnu/bin/ranlib
STRIP=/usr/x86_64-linux-gnu/bin/strip
Configuration (profile_build):
[settings]
arch=armv8
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8.3
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.txt: Installing package
Requirements
    zlib/1.2.12 from 'conancenter' - Cache
Packages
    zlib/1.2.12:9f5d845a9acd4104ba44865c80f5b042812f136e - Build
Build requirements
    zlib/1.2.12 from 'conancenter' - Cache
Build requirements packages
    zlib/1.2.12:1cfd15a0464c4b1c680c61895450fbdbe189331c - Cache

Cross-build from 'Linux:armv8' to 'Linux:x86_64'
Installing (downloading, building) binaries...
zlib/1.2.12: Already installed!
zlib/1.2.12: Copying sources to build folder
zlib/1.2.12: Building your package in /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e
zlib/1.2.12: Generator txt created conanbuildinfo.txt
zlib/1.2.12: Calling generate()
zlib/1.2.12: Aggregating env generators
zlib/1.2.12: Calling build()
zlib/1.2.12: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/src"
-- Using Conan toolchain: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is GNU 8.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/x86_64-linux-gnu-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - done
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Renaming
--     /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/src/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/Release
zlib/1.2.12: CMake command: cmake --build "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/Release" '--' '-j6'
Scanning dependencies of target zlib
[  6%] Building C object CMakeFiles/zlib.dir/adler32.c.o
[ 18%] Building C object CMakeFiles/zlib.dir/crc32.c.o
[ 18%] Building C object CMakeFiles/zlib.dir/deflate.c.o
[ 25%] Building C object CMakeFiles/zlib.dir/compress.c.o
[ 31%] Building C object CMakeFiles/zlib.dir/gzlib.c.o
[ 37%] Building C object CMakeFiles/zlib.dir/gzclose.c.o
[ 43%] Building C object CMakeFiles/zlib.dir/gzwrite.c.o
[ 50%] Building C object CMakeFiles/zlib.dir/gzread.c.o
[ 56%] Building C object CMakeFiles/zlib.dir/inflate.c.o
[ 62%] Building C object CMakeFiles/zlib.dir/infback.c.o
[ 68%] Building C object CMakeFiles/zlib.dir/inftrees.c.o
[ 75%] Building C object CMakeFiles/zlib.dir/inffast.c.o
[ 81%] Building C object CMakeFiles/zlib.dir/trees.c.o
[ 87%] Building C object CMakeFiles/zlib.dir/uncompr.c.o
[ 93%] Building C object CMakeFiles/zlib.dir/zutil.c.o
[100%] Linking C static library libz.a
[100%] Built target zlib
zlib/1.2.12: Package '9f5d845a9acd4104ba44865c80f5b042812f136e' built
zlib/1.2.12: Build folder /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/Release
zlib/1.2.12: Generated conaninfo.txt
zlib/1.2.12: Generated conanbuildinfo.txt
zlib/1.2.12: Generating the package
zlib/1.2.12: Package folder /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e
zlib/1.2.12: Calling package()
zlib/1.2.12: CMake command: cmake --install "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e/build/Release" --prefix "/home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e"
-- Install configuration: "Release"
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e/lib/libz.a
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e/include/zconf.h
-- Installing: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/package/9f5d845a9acd4104ba44865c80f5b042812f136e/include/zlib.h
zlib/1.2.12 package(): Packaged 1 file: LICENSE
zlib/1.2.12 package(): Packaged 2 '.h' files: zconf.h, zlib.h
zlib/1.2.12 package(): Packaged 1 '.a' file: libz.a
zlib/1.2.12: Package '9f5d845a9acd4104ba44865c80f5b042812f136e' created
zlib/1.2.12: Created package revision 2ebbb2a064abcbe9ff85990aa07e34c2
conanfile.txt: Applying build-requirement: zlib/1.2.12
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Aggregating env generators
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo
zlib/1.2.12: WARN: Package binary is corrupted, removing: 9f5d845a9acd4104ba44865c80f5b042812f136e
zlib/1.2.12: WARN: Build folder is dirty, removing it: /home/developer/Scratch/bug-report/.conan/data/zlib/1.2.12/_/_/build/9f5d845a9acd4104ba44865c80f5b042812f136e
```
</details>

**Comment by jcar87** [comment]
Hi @je894591 - thanks so much for the bug report, and for the detailed steps to reproduce it. We have managed to reproduce this and have diagnosed a potential cause. Will update this issue once there are updates.

**Comment by je894591** [comment]
Thanks for the quick fix!

**Comment by jcar87** [comment]
Hi @je894591  - thank you for reporting this! This fix will be part of the upcoming 1.52 release.
